src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.XmlSnippetParsing.cs

```diff
@@ -0,0 +1,288 @@
+        {
+            return ReadCodeSnippetElements(document).Select(element => new CodeSnippet(element)).ToImmutableArray();
+        }
+    }
+
+    /// <summary>
+    /// Shamelessly adapted from https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Editor/VisualStudio/Impl/Snippet/ExpansionTemplate.cs
+    /// with changes to parsing to store the snippet as a set of parts instead of a single string.
+    /// </summary>
+    private class ExpansionTemplate
+    {
+        private record ExpansionField(string ID, string Default, string? FunctionName, string? FunctionParam, bool IsEditable);
+
+        private const string _selected = "selected";
+        private const string _end = "end";
+
+        private readonly List<ExpansionField> _tokens = new();
+
+        private readonly string? _code;
+        private readonly string _delimiter = "$";
+
+        public ExpansionTemplate(CodeSnippet snippet)
+        {
+            var snippetElement = CodeSnippet.GetElementWithoutNamespace(snippet.CodeSnippetElement, "Snippet");
+            var declarationsElement = CodeSnippet.GetElementWithoutNamespace(snippetElement, "Declarations");
+            ReadDeclarations(declarationsElement);
+            var code = CodeSnippet.GetElementWithoutNamespace(snippetElement, "Code");
+            if (code == null)
+            {
+                throw new InvalidOperationException("snippet is missing code element.");
+            }
+
+            _code = Regex.Replace(code.Value, "(?<!\r)\n", "\r\n");
+            var delimiterAttribute = code.Attributes().FirstOrDefault(a => a.Name.LocalName.Equals("Delimiter", StringComparison.OrdinalIgnoreCase));
+            if (delimiterAttribute != null)
+            {
+                _delimiter = delimiterAttribute.Value;
+            }
+        }
+
+        private void ReadDeclarations(XElement? declarations)
+        {
+            if (declarations == null)
+            {
+                return;
+            }
+
+            foreach (var declarationElement in declarations.Elements())
+            {
+                var editableAttribute = declarationElement.Attribute("Editable");
+                var functionElement = CodeSnippet.GetElementWithoutNamespace(declarationElement, "Function");
+                SnippetFunctionService.TryGetSnippetFunctionInfo(functionElement?.Value, out var functionName, out var functionParam);
+                _tokens.Add(new ExpansionField(
+                    CodeSnippet.GetElementInnerText(declarationElement, "ID"),
+                    CodeSnippet.GetElementInnerText(declarationElement, "Default") ?? " ",
+                    functionName,
+                    functionParam,
+                    editableAttribute == null || string.Equals(editableAttribute.Value, "true", StringComparison.Ordinal) || string.Equals(editableAttribute.Value, "1", StringComparison.Ordinal)));
+            }
+        }
+
+        public ParsedXmlSnippet? Parse()
```

Comment:
much of the parsing logic remains the same as the battle-tested editor code, the major change is to convert the snippet into 'SnippetParts' instead of just a pure string with default values.